### ISSUE TITLE
Sites table is completely unusable with a screen reader

### ISSUE THREAD

**Issue by Simon818** [2factorauth/frontend#37]
At the moment it is not possible for me to tell which 2FA options are available on each listed site, and navigation is difficult. This can be broken down into two related issues:

1. The list of sites is not actually a table.
2. The symbols used for indicating a 2FA option are not read by screen readers.

Screen readers can navigate a proper HTML table in any of the four directions, and with valid table row/column headers, everything should be read properly. Unfortunately right now, the lack of a real table means that I don't get any spatial information and can't tell which options are available for every site. I can probably download the files and read them that way, but it would be great if the public-facing site were also usable.

As for subissue 2, since row/column headers are readable, it would be helpful if the 2FA cells had a "yes" or "no" aria-label. The screen reader will fill in the missing information, prefixing the "yes" or the "no" with the appropriate header depending on the direction of movement. There's no need to include the name of the service or the 2FA type in the aria-label.

**Comment by Carlgo11** [comment]
I think the mobile layout works better with screen readers so ideally we could serve that to screen readers. Unsure how feasible that is though

**Comment by Simon818** [comment]
You're right, that is a lot better. There's no way to automatically serve that to screen reader users though, as not all platforms offer a way for a website to determine that a screen reader is loaded, and it's possible to turn this setting off (which I would do for privacy reasons.)

**Comment by Carlgo11** [comment]
Thanks for the input @Simon818!

After researching for a couple of hours I can only conclude that screen reader support on the internet is lacking, to say the least.
I did manage to fix some issues with how the content is displayed but it seems nearly impossible to serve the crucial data, i.e. what 2FA methods a site supports.

The reason we moved away from the table layout was that the file size of the pages was getting out of hand as the tables needed to be generated both in a desktop version and a mobile version. Using our current CSS Grid layout we're able to just use one layout and with CSS render dynamically based on the screen size of the client. As a result, we're now able to serve people with limited internet connections.
Unfortunately, as you've pointed out, it means less non-graphical context. In the best of world, we would be able to serve screen readers appropriate data based on CSS media queries but it seems that's not an option as `media aural` [support is now nonexistent](https://caniuse.com/sr_css-aural).

<hr>

I'm adding the following table of ARIA issues on the site and possible fixes so that anyone who wants to contribute in making the site more accessible can see what needs to be worked on.

|Issue|Solution|
|--------|------------|
| Show 2FA methods on Desktop | No CSS media support for screen readers to allow display `.tfa-summary` only to screen readers (alongside mobile devices). |
| ARIA hide bootstrap icons | Not fixable due to lack of [browser support](https://caniuse.com/mdn-css_properties_content_alt_text). |
| Associate table with category button | Add `aria-labelledby` for each table. |
| Focus on table after category button click | No solution found as `autofocus` breaks focus for non-screen readers. |
| Skip straight to `<main>` | Fixable with `tabindex -1` button. |

**Comment by Simon818** [comment]
Thanks for looking into this. I admit I'm not an expert and the lack of support for some of this comes as a surprise.

I wonder if the first issue is solvable with ARIA, or maybe even hidden text in something like an sr-only class, which adds text to the page but hides it visually. This is basically the opposite of aria-hidden. Or maybe it would be possible to make a span or other control with an aria-label. I'd need to experiment with this a bit. It's not the end of the world if the icons are exposed to the screen reader, because the characters don't get announced anyway.

In terms of focusing the table after categories are selected, this may not be doable but I think a heading before the table and categories list would be helpful, or failing that, different landmarks for each. Right now I take advantage of there being no links in the categories list and just jump to the next link, which will always be the first site in the chosen category. However, you're right that best practice would be either focusing the list or announcing that new content had loaded. Could the categories be radio buttons or toggle buttons with [aria-pressed](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-pressed) set appropriately on them? This would cause the screen reader to announce "pressed" or "checked" after selecting one of them, which would clue the user into the fact that something had happened.

Either way, I really appreciate you spending this much time looking into it; that's kind of rare. If I'm throwing nonsense ideas out, feel free to disregard.

**Comment by Carlgo11** [comment]
Hi again,
we recently pushed some changes to the HTML structure that should improve readability for screen readers.
Please let me know if these changes fixed the issues for you. 🙂

**Comment by Simon818** [comment]
Hi,

Any reason you set aria-hidden on the "title" divs? It means the names of the sites are not reported. So I see a bunch of "Documentation" links and 2FA methods with no idea what they are associated with--basically, the opposite problem. I see you aria-labelled the outer div containers with the site's name, but because divs are pretty generic by default, that isn't going to be reported to most screen readers unless you use some kind of HTML tag more meaningful than div (or give them the equivalent role). <article> would work well for this, actually. I'm sure it's not quite what the spec had in mind but it'll do. Even if you do that, I think it's important to unhide the title div because otherwise screen reader users can't click the link to get to the site or read its name properly in the context of the rest of the page. Aria-hidden is dangerous if used even slightly incorrectly.

But other than that, it looks like your fix worked, so thank you. I removed aria-hidden on everything and I'm able to read the sites list. I think the only missing piece is the ability to jump straight from the sections to the list of 2FA sites (could each of those be a [region](https://www.w3.org/WAI/tutorials/page-structure/regions/)?) and from one 2FA site entry to the next, which should be solvable with the article tag suggestion.

**Comment by Carlgo11** [comment]
Thanks for the feedback!
One step forward two steps back, right? 
For my next repertoire, I'll now proceed to turn the entire page into an image. :smile:

The entry name is an easy fix. I'm less familiar with the regions. Right now the page is divided into three "landmarks". A navbar at the top, the main content, and finally the footer. They all use appropriate tags such as `nav`, `main`, and `footer` as recommended in the [page regions](https://www.w3.org/WAI/tutorials/page-structure/regions/) documentation you sent. Are you not able to jump straight to the main element?

Regarding the ability to jump between entries, would a role of "row" for each entry be sufficient?

**Comment by Simon818** [comment]
> For my next repertoire, I'll now proceed to turn the entire page into an image. 😄

Reminds me of the old sites that used Flash for literally the entire interface. That was a welcome death for the blind community.😄

> Are you not able to jump straight to the main element?

I can, but that still leaves me unable to easily jump from the list of categories to the list of sites after they've loaded, and back again.

> would a role of "row" for each entry be sufficient?

I think that only works if you have the surrounding ARIA to construct a "fake" table. Rows are navigable within a table, but if there is no table, the screen reader will just fail to move anywhere. I recommended \<article\> because there's no need for any additional HTML to support it; it just works, kind of like \<h1\> etc. But it also (as far as I know) doesn't add any additional formatting. It's just a fancy div that has semantic meaning. You could probably also add `role="article"` to accomplish the same thing if I'm wrong about it not creating visual changes.

Oh, and thanks for that change; it works great now.

**Comment by Simon818** [comment]
Hi,

Your development site does automatically move the focus to the table after selecting a category, which is nice. I don't think it's necessary to successfully use the site, but it would be a welcome addition if you can get it to work. However, I think the biggest issue is still the lack of any meaningful separation between the categories list and the sites table. Screen reader users almost never navigate with the tab key because we have so many other commands to move to the next or previous link/heading/button/article/region/landmark/just-about-anything-else-you-can-think-of, but we don't have a way to move to the next \<div\> that has nothing special about it. This is why I suggested using a heading and/or regions to separate the two main sections of the page, and having each site have a role of article. If that isn't easy to do in the generator, we can keep thinking, but I realized I never explained why I thought any of that was a good idea. A well marked up page can literally be 50 times more efficient to navigate because all that markup means users can instantly jump around on the page, whereas a page with a bunch of divs and spans and no ARIA will force me to just press the arrow keys over and over.

If you really wanted to make this site navigable with only the keyboard, you could. You'd probably want to just remove the tabindex from all the buttons and add it to the outer list, then have it listen for arrow key presses and have it move the user's focus from one category button to the next. Then you do the same with the sites table, so users can easily just tab between the categories and the sites. But I'm not as concerned about that for screen reader navigation, because every screen reader has commands or gestures to move to the next heading and landmark, and not every device even has a keyboard.

**Comment by Carlgo11** [comment]
I'm afraid it's not possible to separate the category buttons from the table(s) as they need to have the same parent element for the CSS Grid to work. If I surround the category buttons with a `<section>` or similar, it won't be possible to reposition the different elements as required for the mobile and desktop views to function properly.

The best I can do with the HTML structure is to generate the table after its' category button instead of the current model where all category buttons are generated and then all tables are generated. Would that be of any help?

**Comment by Simon818** [comment]
I think you misunderstand what I mean by "separating." What I'm trying to accomplish is just adding the proper metadata into the site so the screen reader understands what's going on enough to jump around the page. I see you've done this already (sort of), by adding `role="row"` to each entry and `role="table"` to each list of sites. However, table navigation seems to break in most screen readers unless you also have `role="cell"` on each site. Since the table isn't navigable in a traditional way, it doesn't really make sense to "fake" the table structure unless you're interested in constructing the entire thing cell by cell, including adding the 2FA methods as separate cells and keeping the column format exactly consistent; and that would require quite a rework for little gain and basically brings us back to square 1. So, regions and articles were my next best idea.

Just to test this, I ran the following code on the page and it resulted in exactly what I had in mind:

```
for (let el of document.querySelectorAll(".table")) el.setAttribute("role", "region");
for (let el of document.querySelectorAll(".entry")) el.setAttribute("role", "article");
```

This enables the "next region" hotkey to jump straight to the list of sites and the "next article" hotkey to jump from one site to the next. There's a heading before the list of categories and it's also a \<main\>, so getting back there is fairly easy. As far as I know, changing the role doesn't upset the layout at all, though I can't verify this. Am I missing something?

**Comment by Carlgo11** [comment]
Oh okay! I updated the development page to reflect your suggested changes.
Let me know if it works as expected for you.

**Comment by Simon818** [comment]
@Carlgo11 Apologies for the delay in getting back to this; I see it was closed today but I also had it on my to-do list to comment today. It looks like this works exactly as expected and it's an absolute delight to navigate now. Thanks so much for being receptive and working on this.
